### PR TITLE
Implementation of call functions on iteration state

### PR DESCRIPTION
**Pull request by zloidemon**
* Added the ability to set iteration and full scan delays for a task.
* Added callbacks for a task at various stages of the full scan iteration.

@ChangeLog:
* Added the ability to set iteration and full scan delays for a task.
* Added callbacks for a task at various stages of the full scan iteration.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -44,6 +44,11 @@ opt
   * `args` - passed to `is_tuple_expired()` and `process_expired_tuple()` as additional context
   * `tuples_per_iteration` - number of tuples will be checked by one iteration
   * `full_scan_time` - time required for full index scan (in seconds)
+  * `on_iteration_error` - call function on error
+  * `on_iteration_start` - call function on iteration
+  * `on_iteration_complete` - call function on complete iteration
+  * `on_iteration_stop` - call function on end iteration cycle
+  * `max_delay` - delay time between iterations
```
Comment (Totktonada) [inline README.md:51]: I would split `max_delay` to `tuples_per_iteration_delay` and `full_scan_delay`. And I would provide a constant like `expirationd.DELAY_INFINITY = 500 * 365 * 86400 -- 500 years` to pass to `tuples_per_iteration_delay`, when one want to discard the upper limit.

File: README.md [hunk 2]
```diff
@@ -44,6 +44,11 @@ opt
   * `args` - passed to `is_tuple_expired()` and `process_expired_tuple()` as additional context
   * `tuples_per_iteration` - number of tuples will be checked by one iteration
   * `full_scan_time` - time required for full index scan (in seconds)
+  * `on_iteration_error` - call function on error
+  * `on_iteration_start` - call function on iteration
+  * `on_iteration_complete` - call function on complete iteration
+  * `on_iteration_stop` - call function on end iteration cycle
```
Comment (Totktonada) [inline README.md:50]: The terminology here is confusing (just like in the code itself). Those callbacks are invoked after the whole space traversal and the traverse / full scan is called 'iteration'. But `tuples_per_iteration` assumes 'iteration' relation to a chunk of tuples of given size. Don't know what to do with this naming...

Most logical would be name `tuples_per_iteration` as `chunk_size`. However, it is named so for years and renaming may lead to even more confusion. So, maybe, it worth to left the 'iteration' term to the chunk of tuples and name the new options as `on_traverse_*` or `on_full_scan_*`? The latter is consistent with `full_scan_time` that looks as the good property.

**Comment by Totktonada** [comment]
@zloidemon The changes you propose looks useful for me. I would rebase the PR on top of the current master, implement the changes I proposed above and split the commit into two ones: one re delays and another re callbacks. Aren't you mind?

**Comment by Totktonada** [comment]
(Sorry for the looong delay.)

File: README.md (hunk 1 above)
Comment (LeonidVas) [reply README.md]: Done.

Comment (LeonidVas) [reply README.md]: expirationd.DELAY_INFINITY doesn't make sense. This delay will hang the fiber forever.

File: README.md (hunk 2 above)
Comment (LeonidVas) [reply README.md]: Done.

File: README.md [hunk 3]
```diff
@@ -44,6 +44,16 @@ opt
...
+  * `iteration_delay` - max sleep time between iterations (in seconds)
+  * `full_scan_delay` - sleep time between full scans (in seconds)
+  * `on_full_scan_start` - call function on starting full scan iteration.
+    Receives `(task)` as arguments.
+  * `on_full_scan_complete` - call function on complete full scan iteration
+    Receives `(task)` as arguments.
+  * `on_full_scan_success` - call function on success full scan iteration
+    Receives `(task)` as arguments.
+  * `on_full_scan_error` - call function on error full scan iteration
+    Receives `(task, error)` as arguments.
```
Comment (artemreyt) [inline README.md:56]: For me, it isn't obvious that `on_full_scan_complete` will be called after `on_full_scan_success` or `on_full_scan_error`. I propose to mention it.

**Review by artemreyt — COMMENTED** [review]
Also, there isn't test for `on_full_scan_error`.  Maybe add it?

File: README.md (hunk 3 above)
Comment (LeonidVas) [reply README.md:57]: Done.

**Comment by LeonidVas** [comment]
> Also, there isn't test for `on_full_scan_error`. Maybe add it?

Added.

**Review by artemreyt — COMMENTED** [review]
LGTM.

**Review by artemreyt — APPROVED** [review]
LGTM.

**Pull request merged**